polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -1256,31 +1333,369 @@ impl<T: Config> Pallet<T> {
+					]
+					.into_iter(),
+				)
+				.collect(),
+		);
+		Ok((local_execute_xcm, xcm_on_dest))
+	}
+
+	// function assumes fees and assets have the same remote reserve
+	fn remote_reserve_transfer_message(
+		reserve: MultiLocation,
+		dest: MultiLocation,
+		beneficiary: MultiLocation,
+		assets: Vec<MultiAsset>,
+		mut fees: MultiAsset,
+		weight_limit: WeightLimit,
+	) -> Result<Xcm<<T as Config>::RuntimeCall>, Error<T>> {
+		let max_assets = assets.len() as u32;
+		let context = T::UniversalLocation::get();
+		// we spend up to half of fees for execution on reserve and other half for execution on
+		// destination
+		match &mut fees.fun {
+			Fungible(amount) => *amount = amount.saturating_div(2),
+			NonFungible(_) => return Err(Error::<T>::FeesNotMet),
+		};
+		// identifies fee item as seen by `reserve` - to be used at reserve chain
+		let reserve_fees = fees
+			.clone()
+			.reanchored(&reserve, context)
+			.map_err(|_| Error::<T>::CannotReanchor)?;
+		// identifies fee item as seen by `dest` - to be used at destination chain
+		let dest_fees = fees.reanchored(&dest, context).map_err(|_| Error::<T>::CannotReanchor)?;
+		// identifies `dest` as seen by `reserve`
+		let dest = dest.reanchored(&reserve, context).map_err(|_| Error::<T>::CannotReanchor)?;
+		// xcm to be executed at dest
+		let xcm_on_dest = Xcm(vec![
+			BuyExecution { fees: dest_fees, weight_limit: weight_limit.clone() },
+			DepositAsset { assets: Wild(AllCounted(max_assets)), beneficiary },
+		]);
+		// xcm to be executed on reserve
+		let xcm_on_reserve = Xcm(vec![
+			BuyExecution { fees: reserve_fees, weight_limit },
+			DepositReserveAsset { assets: Wild(AllCounted(max_assets)), dest, xcm: xcm_on_dest },
+		]);
+		Ok(Xcm(vec![
+			WithdrawAsset(assets.into()),
+			InitiateReserveWithdraw {
+				assets: Wild(AllCounted(max_assets)),
+				reserve,
+				xcm: xcm_on_reserve,
+			},
+		]))
+	}
+
+	fn prefund_teleport_fees_messages(
+		dest: MultiLocation,
+		fees: MultiAsset,
+	) -> Result<(Xcm<<T as Config>::RuntimeCall>, Xcm<()>), Error<T>> {
+		let context = T::UniversalLocation::get();
+		let reanchored_fees = fees
+			.clone()
+			.reanchored(&dest, context)
+			.map_err(|_| Error::<T>::CannotReanchor)?;
+
+		// XcmContext irrelevant in teleports checks
+		let dummy_context =
+			XcmContext { origin: None, message_id: Default::default(), topic: None };
+		// We should check that the asset can actually be teleported out (for this to
+		// be in error, there would need to be an accounting violation by ourselves,
+		// so it's unlikely, but we don't want to allow that kind of bug to leak into
+		// a trusted chain.
+		<T::XcmExecutor as XcmAssetTransfers>::AssetTransactor::can_check_out(
+			&dest,
+			&fees,
+			&dummy_context,
+		)
+		.map_err(|_| Error::<T>::CannotCheckOutTeleport)?;
+		<T::XcmExecutor as XcmAssetTransfers>::AssetTransactor::check_out(
+			&dest,
+			&fees,
+			&dummy_context,
+		);
```

Comment:
I am just thinking if this shouldnt be better done in `XcmExecutor` e.g. maybe with some XcmV4 new instruction,
or maybe if we could add new function `XcmAssetTransfers::check_out_assets` and then refactor `XcmExecutor` [this](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-executor/src/lib.rs#L655-L664) to the inner function which would be also called by `XcmAssetTransfers::check_out_assets`, 

this way we can do `AssetTransactor::check_out` just on one place and also we dont need to expose `AssetTransactor` with `XcmAssetTransfers`

Reply:
hmm.. to do this properly, we'd have to pull out all usages of `Config::AssetTransactor::*` functions from `XcmExecutor::process_instruction()` implementation to `XcmAssetTransfers::*` helper functions - it'd be inconsistent/incomplete to do it just for `AssetTransactor::check_out`..

as that's a non-negligible XcmExecutor change/refactor, I would definitely do it in a follow-up PR.